File: p0017.py
"""
Number letter counts
Problem 17

https://projecteuler.net/problem=17
"""


def sol():
    under_ten = get_under_ten()
    total_of_one_to_ninety_nine = get_total_of_one_to_ninety_nine(under_ten)
    total_of_one_hundred_until_thousand = get_total_of_one_hundred_until_thousand(
        under_ten, total_of_one_to_ninety_nine
    )
    print(
        total_of_one_to_ninety_nine
        + total_of_one_hundred_until_thousand
        + len("one")
        + len("thousand")
    )


def get_under_ten():
    return [
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
    ]


def get_total_of_one_to_ninety_nine(under_ten):
    ten_to_nineteen = [
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen",
    ]
    twenty_to_ninety_step_10 = [
        "twenty",
        "thirty",
        "forty",
        "fifty",
        "sixty",
        "seventy",
        "eighty",
        "ninety",
    ]
    total_length_of_under_ten = sum([len(i) for i in under_ten])
    total_length_of_ten_to_nineteen = sum([len(i) for i in ten_to_nineteen])
    total_length_of_twenty_to_ninety_nine = 0
    for ten_time in twenty_to_ninety_step_10:
        tmp = len(ten_time) * 10
        tmp += total_length_of_under_ten
        total_length_of_twenty_to_ninety_nine += tmp
    return (
        total_length_of_under_ten
        + total_length_of_ten_to_nineteen
        + total_length_of_twenty_to_ninety_nine
    )


def get_total_of_one_hundred_until_thousand(under_ten, total_of_one_to_ninety_nine):
    result = 0
    hundred_length = len("hundred")
    and_length = len("and")
    for num in under_ten:
        tmp = (len(num) + len("hundred")) * 100
        tmp += and_length * 99
        tmp += total_of_one_to_ninety_nine
        result += tmp
    return result


if __name__ == "__main__":
    sol()

File: elixir/p0017.ex
defmodule P0017 do
  @moduledoc """
  Number Letter Counts
  Problem 17

  https://projecteuler.net/problem=17
  """

  def solve() do
    1..1000
    |> Enum.map(fn x -> number_to_words_until_thousand(x) end)
    |> Enum.map(fn x -> String.length(x) end)
    |> Enum.sum()
  end

  def number_to_words_until_thousand(number) when number <= 0 do
    ""
  end

  def number_to_words_until_thousand(number) when 1 <= number and number < 10 do
    first_digits = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
    Enum.at(first_digits, number)
  end

  def number_to_words_until_thousand(number) when 10 <= number and number < 20 do
    ten_to_nineteen = [
      "ten",
      "eleven",
      "twelve",
      "thirteen",
      "fourteen",
      "fifteen",
      "sixteen",
      "seventeen",
      "eighteen",
      "nineteen"
    ]

    Enum.at(ten_to_nineteen, number - 10)
  end

  def number_to_words_until_thousand(number) when 20 <= number and number < 100 do
    tens = ["", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]

    Enum.at(tens, div(number, 10)) <>
      number_to_words_until_thousand(number - div(number, 10) * 10)
  end

  def number_to_words_until_thousand(number) when 100 <= number and number < 1000 do
    first_digits = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

    if rem(number, 100) == 0 do
      Enum.at(first_digits, div(number, 100)) <> "hundred"
    else
      Enum.at(first_digits, div(number, 100)) <>
        "hundred" <>
        "and" <>
        number_to_words_until_thousand(number - div(number, 100) * 100)
    end
  end

  def number_to_words_until_thousand(number) when number == 1000 do
    "one" <> "thousand"
  end
end

P0017.solve()
|> IO.puts()
